go-controller/pkg/network-attach-def-controller/network_attach_def_controller.go

```diff
@@ -503,6 +505,20 @@ func (nadController *NetAttachDefinitionController) addNADToController(ncm Netwo
 						nadController.name, nadName, networkName)
 				}
 			}
+			if nni.nadInDeletion != "" {
+				if nni.nadInDeletion == nadName {
+					// the NAD being added is exactly the same as NAD being deleted, keep trying to delete the old NAD first
+					err = nadController.deleteNADFromControllerCommon(nni, networkName, nadName)
+					if err != nil {
+						return err
+					}
+					// if the previous NAD is deleted successfully, return error here, so the new NAD will be added later
+					return fmt.Errorf("%s: previous NAD %s in network %s in deletion is now deleted successfully, "+
```

Comment:
this seems very icky that you are banging on the previous NAD's deletion which completed successfully to error out for the new NAD's creation to be retried?
for the default network controller, the retry framework takes care of this logic isn't there something similar we can do?

long term: this really sounds like we need level driven controllers for NADs

Reply:
this is the level driven controller, not the same as retry framework for default network controller. The thing is level driven is based on key name (here the NAD name), and that is why we need to clean up the old NAD even this function is driven by a different NAD newly created but with the same key.

Reply:
why cant you just continue to add the nad back in this function?

Reply:
because the dead controller is not deleted yet, and this PR is to avoid reusing the same controllre for the new NAD if the previous NAD deletion failed.

Reply:
we cannot, we have to return from `addNADToController()` first, so that we don't reuse the old controller. again, the old controller cannot be reused as it is already stopped/cleaned-up, for example, its stopChan is already stopped.

Reply:
maybe using uuid as the key of level driven as @maiqueb suggested is a good idea.